app/controllers/register/establishers/company/CompanyAddressListController.scala

```diff
@@ -16,73 +16,78 @@
-      }
+class CompanyAddressListController @Inject()(
+        override val appConfig: FrontendAppConfig,
+        override val cacheConnector: DataCacheConnector,
+        override val navigator: Navigator,
+        override val messagesApi: MessagesApi,
+        authenticate: AuthAction,
+        getData: DataRetrievalAction,
+        requireData: DataRequiredAction) extends AddressListController {
+
+  def onPageLoad(mode: Mode, index: Index): Action[AnyContent] = (authenticate andThen getData andThen requireData).async { implicit request =>
+
+    addressListViewModel(mode, index) match {
+      case Some(viewModel) => get(viewModel)
+      case _ => redirectToPostCodeLookup(mode, index)
+    }
+
+  }
+
+  def onSubmit(mode: Mode, index: Index): Action[AnyContent] = (authenticate andThen getData andThen requireData).async { implicit request =>
+
+    addressListViewModel(mode, index) match {
+      case Some(viewModel) => post(viewModel, CompanyAddressId(index), mode)
+      case _ => redirectToPostCodeLookup(mode, index)
+    }
+
   }
 
-  def onSubmit(mode: Mode, index: Index): Action[AnyContent] = (authenticate andThen getData andThen requireData).async {
-    implicit request =>
-      retrieve(mode, index) {
-        case (name, addresses) =>
-          formProvider(addresses).bindFromRequest().fold(
-            form =>
-              Future.successful(BadRequest(
-                companyAddressList(appConfig, form, mode, index, addresses, name)
-              )),
-            id =>
-              dataCacheConnector.save(request.externalId, CompanyAddressId(index), addresses(id).copy(country = "GB")).map {
-                json =>
-                  Redirect(navigator.nextPage(CompanyAddressListId(index), mode)(UserAnswers(json)))
-              }
-          )
-      }
+  private def addressListViewModel(mode: Mode, index: Index)(implicit request: DataRequest[AnyContent]): Option[AddressListViewModel] = {
+
+    retrieve(index).map { case (companyName, addresses) =>
+        val postCall = routes.CompanyAddressListController.onSubmit(mode, index)
+        val manualInputCall = routes.CompanyAddressController.onPageLoad(mode, index)
+
+        AddressListViewModel(
+          postCall,
+          manualInputCall,
+          addresses,
+          subHeading = Some(Message(companyName))
+        )
+    }
+
   }
 
-  private def retrieve(mode: Mode, index: Int)(f: (String, Seq[Address]) => Future[Result])
-                      (implicit request: DataRequest[AnyContent]): Future[Result] = {
-    retrieveCompanyName(index){ companyName =>
-      request.userAnswers.get(CompanyPostCodeLookupId(index)).map { addresses =>
-        f(companyName, addresses)
-      } getOrElse {
-        Future.successful(Redirect(controllers.register.establishers.company.routes.CompanyPostCodeLookupController.onPageLoad(mode, index)))
-      }
+  private def retrieve(index: Index)(implicit request: DataRequest[AnyContent]): Option[(String, Seq[Address])] = {
```

Comment:
This could be done with the `Retrievals` syntax stuff,

```scala
(CompanyDetailsId(index) and CompanyPostCodeLookupId(index)).retrieve.right.map {
  case companyDetails ~ postcodeLookup =>
    ???
}
```